### PR TITLE
Python 3 updates

### PR DESCRIPTION
**Pull request by borrob**
Please do *NOT* approve this pull request. This is Work In Progress for the python3 conversion. I have an issue with one specific test and we want to see if Travis gets the same error or if it passes without problems.

### CODE REVIEW

**Comment by justb4** [comment]
The Travis failure is only a `flake8` error.

**Comment by borrob** [comment]
:) yes, so the normal testing passes, so I guess it is something with my setup that the one test fails on my machine.

I haven't yet had much time to work on the docker issue, but otherwise I would say we are almost there.

**Comment by justb4** [comment]
I can helpout with any Docker issues (what was the problem?).

**Comment by borrob** [comment]
I have been able to track the error, but haven't yet able to solve it. I have updated the Dockerfile to use `python3.7.4-alpine3.10` as a starting image. Somewhere in the building process `install.sh` installs the python dependencies from the `requirements.txt` and in there: `pyproj` causes an error:

```
Proj executable not found. Please set PROJ_DIR variable.
```

I tried adding `proj` to the apk installation in the Dockerfile but to no avail. An alternative solution that I found on the [web](https://stackoverflow.com/a/51972021) is to include the python-dev library, but that gets the development files for python 3.7.3, which make me think the virtual env with python `3.7.4` still cannot use it (and I did some other testing suggesting that solely adding python-dev is not enough).

Any ideas or suggestions are welcome. Unfortunately I won't be able to work on this over the weekend.

**Comment by justb4** [comment]
Docker: Think Alpine images are problematic with binaries like GDAL and Proj. May end up doing compiles, or have things improved recently?  Debian Slim is a better option. As with Stetl py3 migration we end up using  `python:3.6-slim-stretch`. And py 3.6 i.s.o. 3.7, can't recall why.
(also busy w other stuff over weekend further...)

**Comment by justb4** [comment]
Sorry, forget my comment on Alpine: was mistaken with other projects (Stetl, pygeoapi, workshop) where GDAL was used. Alpine was already in use with GHC for Python 2.7. I can look into the Docker build further.

**Comment by justb4** [comment]
Globally testing @borrob `py3` branch on 24.09.2019:

* virtualenv with Python 3.7.1 (using `pyenv` on Mac OSX).
* `paver` installation: OK.
* unittests run OK.
* flake8 errors: 
```
(geohealthcheck3.7.1) nusa:borrob.git just$ flake8
./tests/test_plugins.py:41:1: E402 module level import not at top of file
./tests/test_plugins.py:42:1: E402 module level import not at top of file
./tests/test_plugins.py:43:1: E402 module level import not at top of file
./tests/test_plugins.py:44:1: E402 module level import not at top of file
./tests/test_plugins.py:45:1: E402 module level import not at top of file
./tests/test_resources.py:41:1: E402 module level import not at top of file
./tests/test_resources.py:42:1: E402 module level import not at top of file
./tests/test_resources.py:43:1: E402 module level import not at top of file
./tests/test_resources.py:44:1: E402 module level import not at top of file
./tests/test_resources.py:45:1: E402 module level import not at top of file
```
Easily fixed by moving imports to top-level.
* running GHC locally
* adding a WMS:

```
2019-09-24 12:39:33,105 - healthcheck - ERROR - Getting metadata failed: 'str' object has no attribute 'decode'
Traceback (most recent call last):
  File "/Users/just/project/geohealthcheck/borrob.git/GeoHealthCheck/healthcheck.py", line 247, in sniff_test_resource
    title = title.decode('utf-8')
AttributeError: 'str' object has no attribute 'decode'
```

Removing this line [title = title.decode('utf-8')](https://github.com/borrob/GeoHealthCheck/blob/py3/GeoHealthCheck/healthcheck.py#L247) fixes that. (all `str` is unicode in py3 IMHO).

* checking more occurrences of `decode`:  in [models.py](https://github.com/borrob/GeoHealthCheck/blob/py3/GeoHealthCheck/models.py#L699) : token generation for user-initiated password reset.

**Comment by justb4** [comment]
Looked into the `pyproj` Docker build failure: 

* Image based on `python:3.7.4-alpine3.10`
* got indeed error: `Proj executable not found. Please set PROJ_DIR variable.`
* looks similar to earlier [`pyproj` build problem in GHC](https://github.com/geopython/GeoHealthCheck/issues/240) and also [similar Docker build failures](https://github.com/yagajs/docker-mapproxy/issues/13).
* downgraded pyproj to 1.9.6 at least fixes build
* (have to go) to be continued

**Comment by borrob** [comment]
Thanks. I think we are actually almost there. From the top of my head:
- fixing the docker build
- merging the latest changes from master
- fix the flake8 (it might not be as easily as moving the import statements, since the relative/absolue import changed - for now I made a crude solution to be able to run the tests)
- update documentation
- I thought I got all the UTF-8 things - will double check (could also be I haven't pushed the latest changes yet)

**Comment by borrob** [comment]
Docker seems to be working now (by downgrading pyproj), no flake8 issues anymore, all tests are passing, and I merged to latest master branch.

I'll start with reviewing the documentation to see if that is in line with py3.

**Comment by justb4** [comment]
Started testing locally with [py3](https://github.com/borrob/GeoHealthCheck/tree/py3) branch, will add findings in comments:

## 1. all tests run OK

## 2. filter() returns `map` not `list` in Py3
* add a WFS e.g. https://geodata.nationaalgeoregister.nl/aan/wfs
* in Edit try to add any Probe for bbox query 
* see error in UI (cannot add Probe)
* in log
```
  File "/Users/just/project/geohealthcheck/borrob.git/GeoHealthCheck/plugins/probe/wfs.py", line 146, in expand_params
    if len(ft_namespaces) > 0:
TypeError: object of type 'filter' has no len()
```
* solution: in https://github.com/borrob/GeoHealthCheck/blob/py3/GeoHealthCheck/plugins/probe/wfs.py#L142 replace line with `ft_namespaces = list(filter(None, list(ft_namespaces)))`
* reason: [see here](https://stackoverflow.com/questions/41903852/typeerror-object-of-type-map-has-no-len-python3)
* will look for more `filter()` occurrences assuming `list`...

**Comment by justb4** [comment]
Tested after pulling above commits.

## 5 HttpStatusNoError Never detects errors
Was testing Basic Auth. Curious why even without credentials `HTTPGet Probe` was successful. We trapped into the *div issue* from Python2 to Python3! Problem is [this line in checks.py](https://github.com/geopython/GeoHealthCheck/blob/master/GeoHealthCheck/plugins/check/checks.py#L28):
```
    def perform(self):
        """Default check: Resource should at least give no error"""
        status = self.probe.response.status_code
==>     overall_status = status / 100
        if overall_status in [4, 5]:
            self.set_result(False, 'HTTP Error status=%d' % status)
```
In Python2 gives `floor` value, but Python3 a `float`. 
Fix needs to be: `overall_status = status // 100` AFAICS.
(Now Basic Auth always fails, even with right credentials, but is other (encoding?) issue...)

**Comment by justb4** [comment]
Ok, tested again: 3 (OWSLib TMS) and 5 (HTTP status check) above found solved. 4 (auth encoding) only with SQLite as DB.

## 6 - Auth info Decoding/Encoding Failure with Postgres
Only occurs when using Postgres, not with SQLite. Background: GHC encodes/encrypts a generic auth dict info structure via JSON string to be stored as textfield in DB. It decodes/decrypts when reading. This way we can support multiple auth types with a single `auth` column in `resource` table. So this is another encoding than in 4 for HTTP auth headers, but think similar problem.
We also need to deal with existing PG DBs that have `auth` columns already present in `resource ` table.
 
* install `psycopg2` : `pip install psycopg2`
* used existing PG DB, only changed in `config_site.py`:  `SQLALCHEMY_DATABASE_URI = 'postgresql://name:passw@localhost:5432/ghc'`, but may create new
* add URL Resource with Basic Auth (even does not have to have basic auth)
* in Edit add Basic Auth Username and Password
* click Save
* Exception: 
```
  File "/Users/just/project/geohealthcheck/borrob.git/GeoHealthCheck/models.py", line 584, in auth_type
    return self.auth['type']
  File "/Users/just/project/geohealthcheck/borrob.git/GeoHealthCheck/models.py", line 578, in auth
    return ResourceAuth.decode(self._auth)
  File "/Users/just/project/geohealthcheck/borrob.git/GeoHealthCheck/resourceauth.py", line 87, in decode
    raise err
  File "/Users/just/project/geohealthcheck/borrob.git/GeoHealthCheck/resourceauth.py", line 83, in decode
    s = decode(APP.config['SECRET_KEY'], encoded)
  File "/Users/just/project/geohealthcheck/borrob.git/GeoHealthCheck/util.py", line 247, in decode
    string = base64.urlsafe_b64decode(string + b'===')
TypeError: can only concatenate str (not "bytes") to str
```

Analysis: problem appears in `decode()` (and probably `encode()` as well) in [util.py](https://github.com/geopython/GeoHealthCheck/blob/master/GeoHealthCheck/util.py#L231) :
```
def decode(key, string):
    string = base64.urlsafe_b64decode(string + b'===')
    string = string.decode('latin') if six.PY3 else string
    encoded_chars = []
    for i in range(len(string)):
        key_c = key[i % len(key)]
        encoded_c = chr((ord(string[i]) - ord(key_c) + 256) % 256)
        encoded_chars.append(encoded_c)
    encoded_string = ''.join(encoded_chars)
    return encoded_string
```

With SQLite `string` is of type `bytes` but with PG of type `string`.
Tried something like:

```
def decode(key, string):
    if type(string) is not bytes:
        string = string.encode()
    string = base64.urlsafe_b64decode(string + b'===')
```

But then get other error on Edit and Test:
```
File "/Users/just/project/geohealthcheck/borrob.git/GeoHealthCheck/util.py", line 247, in decode
string = base64.urlsafe_b64decode(string + b'===')
File "/Users/just/.pyenv/versions/3.7.1/lib/python3.7/base64.py", line 133, in urlsafe_b64decode
return b64decode(s)
File "/Users/just/.pyenv/versions/3.7.1/lib/python3.7/base64.py", line 87, in b64decode
return binascii.a2b_base64(s)
binascii.Error: Invalid base64-encoded string: number of data characters (213) cannot be 1 more than a multiple of 4
```
* don't think `six` is required if we do Py3-only?

**Comment by borrob** [comment]
Weird... I would expect `SQLAlchemy` to deal with the abstraction and get the string/byte conversion right regardless of which database is used. I will look into it.

Agree: we should be able to drop `six`.

**Comment by borrob** [comment]
We're not there yet: I noticed some `paver` commands still need an update because of issues  with the importing of modules. Also: the docker image is building, but I haven't checked yet if it is actually working. I had to change the `gunicorn` configuration on a non-docker deploy.

**Comment by justb4** [comment]
@borrob good to see you synced with `master`! Will try not to do too many big changes. 

The Postgres-char-issue: my bad! For security reasons the `SECRET_KEY` is used for encoding/decoding stored auth creds, and the Py3 instance had a different key! Using the same key: no problem. Pff, was thinking that we had a very serious encoding issue with difficult DB migrations. So we're getting closer!

**Comment by borrob** [comment]
I fixed one paver issue and did some testing (also with docker). I think we're good to go and I'm curious what the results of the demo environment will be.

**Comment by borrob** [comment]
I removed the **draft** tag from this pull request (that was there for the automated testing with travis). Please review and let's hope we can move to py3 soon!

**Comment by justb4** [comment]
Yes, good, I only want to release 0.7.0 from current `master` first, and create a 07.0 maintenance branch. Then further test in particular with Docker and then merge this PR and have a quite some testing time on demo site. OK, @tomkralidis ? Let's aim for all of this before nov 1 ok?

NB solved a nasty concurrency bug (two lines) with #301 #302 today.

File: GeoHealthCheck/healthcheck.py [hunk 1]
```diff
@@ -61,7 +61,7 @@ def db_commit():
     err = None
     try:
         DB.session.commit()
-    except Exception as err:
+    except Exception:
```
Comment (tomkralidis) [inline GeoHealthCheck/healthcheck.py]: While we are here, for consistency, can we log the error, 
```python
            except Exception as err:
                LOGGER.warning("Cannot commit to database %s".format(err))
```

File: GeoHealthCheck/notifications.py [hunk 2]
```diff
@@ -1,5 +1,4 @@
-#!/usr/bin/env python
-# -*- coding: utf-8 -*-
+#!/usr/bin/env python3
```
Comment (tomkralidis) [inline GeoHealthCheck/notifications.py]: Would prefer to remove this shebang line entirely.

File: GeoHealthCheck/util.py [hunk 3]
```diff
@@ -228,22 +227,24 @@ def read(filename, encoding='utf-8'):
 
 
 # https://gist.github.com/gowhari/fea9c559f08a310e5cfd62978bc86a1a
-def encode(key, string):
+# with alterations to get return type of string, not bytes
+def encode(key: str, string: str) -> str:
```
Comment (tomkralidis) [inline GeoHealthCheck/util.py]: Is this annotation specific to Python 3.7+ or safe for 3.4+, say?

**Comment by justb4** [comment]
A great thank-you @borrob! This was no easy PR to do.

**Comment by tomkralidis** [comment]
Thanks for the awesome contribution @borrob!

**Pull request merged**